napari/__main__.py

```diff
@@ -2,32 +2,89 @@
 napari command line viewer.
 """
 import argparse
+import logging
 import sys
+from ast import literal_eval
+from typing import Any, Dict, List
 
 from . import __version__, gui_qt, view_path
 from .utils import citation_text, sys_info
 
 
 class InfoAction(argparse.Action):
     def __call__(self, *args, **kwargs):
+        # prevent unrelated INFO logs when doing "napari --info"
+        logging.basicConfig(level=logging.WARNING)
         print(sys_info())
         sys.exit()
 
 
 class CitationAction(argparse.Action):
     def __call__(self, *args, **kwargs):
+        # prevent unrelated INFO logs when doing "napari --citation"
+        logging.basicConfig(level=logging.WARNING)
         print(citation_text)
         sys.exit()
 
 
+def validate_unknown_args(unknown: List[str]) -> Dict[str, Any]:
+    """Convert a list of strings into a dict of valid kwargs for add_* methods.
+
+    Will exit program if any of the arguments are unrecognized, or are
+    malformed.  Converts string to python type using literal_eval.
+
+    Parameters
+    ----------
+    unknown : List[str]
+        a list of strings gathered as "unknown" arguments in argparse.
+
+    Returns
+    -------
+    kwargs : Dict[str, Any]
+        {key: val} dict suitable for the viewer.add_* methods where ``val``
+        is a ``literal_eval`` result, or string.
+    """
+    from .components.add_layers_mixin import valid_add_kwargs
+
+    out: Dict[str, Any] = dict()
+    valid = set.union(*valid_add_kwargs().values())
+    for i, arg in enumerate(unknown):
+        if not arg.startswith("--"):
+            continue
+        if "=" in arg:
+            sys.exit(f"error: '=' in argument {arg}. (Use space instead)")
```

Comment:
well, it was a bit of a cop out... because it makes parsing with literal_eval much harder.  For instance, these all work:
```
--contrast-limits "(1, 0)"
--contrast-limits "(1,0)"
--contrast-limits="(1,0)"
```
but this does not (because of the space in the tuple)
```
--contrast-limits="(1, 0)"
```

Reply:
I don't understand... use `split('=')` and it won't split on spaces?

```python
import sys
import argparse


parser = argparse.ArgumentParser()

known, unknown = parser.parse_known_args()

print(unknown)
```

```sh
$ python arg.py --contrast-limits="(0, 1)"
['--contrast-limits=(0, 1)']
```

And if I do `print(unknown[0].split('='))`:

```sh
 $ python argp.py --contrast-limits="(0, 1)"
['--contrast-limits', '(0, 1)']
```

Reply:
ah, ok ... your example doesn't have the positional argument with `nargs='*'` (as we do for the `'images'` argument) ... that's what screws it all up:
```bash
# with  parser.add_argument('images', nargs='*', ...
# args, unknown = parser.parse_known_args()
# print(args,unknown)

$ napari path/to/image.tif --contrast-limits "(1, 0)"
Namespace(images=['path/to/image.tif']) ['--contrast-limits', '(1, 0)']

$ napari path/to/image.tif --contrast-limits "(1,0)"
Namespace(images=['path/to/image.tif']) ['--contrast-limits', '(1,0)']

$ napari path/to/image.tif --contrast-limits="(1,0)"
Namespace(images=['path/to/image.tif']) ['--contrast-limits=(1,0)']

$ napari path/to/image.tif --contrast-limits="(1, 0)"
Namespace(images=['path/to/image.tif', '--contrast-limits=(1, 0)']) []  # <- breaks

# with parser.add_argument('images', nargs=1...

$ napari path/to/image.tif --contrast-limits="(1, 0)"
Namespace(images=['path/to/image.tif']) ['--contrast-limits=(1, 0)']
```

Reply:
Damn that's so crap. It should not include things that have a `--`.

Well, this is probably not what you want to hear, but you *could* parse `args.images` and shuffle all the entries that start with `--` over to unknown...?

Reply:
lol... fine.

Reply:
Well, it's that or submit a patch to argparse, your pick... =P

Reply:
fixed!